src/topology/topological_fiber_bundle.lean

```diff
@@ -0,0 +1,398 @@
+trivialization, this construction is defeq to the canonical construction (Σ x : B, F). In the case
+of the tangent bundle of manifolds, this implies that on vector spaces the derivative and the
+manifold derivative are defeq.
+
+A drawback is that some silly constructions will typecheck: in the case of the tangent bundle, one
+can add two vectors in different tangent spaces (as they both are elements of F from the point of
+view of Lean). To solve this, one could mark the tangent space as irreducible, but then one would
+lose the identification of the tangent space to F with F. There is however a big advantage of this
+situation: even if Lean can not check that two basepoints are defeq, it will accept the fact that
+the tangent spaces are the same. For instance, if two maps f and g are locally inverse to each
+other, one can express that the composition of their derivatives is the identity of
+`tangent_space 𝕜 x`. One could fear issues as this composition goes from `tangent_space 𝕜 x` to
+`tangent_space 𝕜 (g (f x))` (which should be the same, but should not be obvious to Lean
+as it does not know that `g (f x) = x`). As these types are the same to Lean (equal to `F`), there
+are in fact no dependent type difficulties here!
+
+For this construction, we should thus choose for each `x` one specific trivialization around it. We
+include this choice in the definition of the fiber bundle, as it makes some constructions more
+functorial and it is a nice way to say that the trivializations cover the whole space B.
+
+With this definition, the type of the fiber bundle is just `Σ x : B, F`, but the topology is not the
+product one.
+
+We also take the indexing type (indexing all the trivializations) as a parameter to the bundle:
+it could always be taken as a subtype of all the maps from open subsets of B to continuous maps
+of F, but in practice it will sometimes be something else. For instance, on a manifold, one will use
+the set of charts as a good parameterization for the trivializations of the tangent bundle. Or for
+the pullback of a fiber bundle the indexing type will be the same as for the initial bundle.
+
+## Tags
+Fiber bundle, topological bundle, vector bundle, local trivialization, structure group
+-/
+
+variables {ι : Type*} {B : Type*} {F : Type*}
+
+open topological_space set
+
+section is_topological_fiber_bundle
+
+variables {Z : Type*} [topological_space B] [topological_space Z]
+  [topological_space F] (proj : Z → B)
+
+variable (F)
+
+/-- A topological fiber bundle with fiber F over a base B is a space projecting on B for which the
+fibers are all homeomorphic to F, such that the local situation around each point is a direct
+product. We do not register the continuity of `proj` in the definition, as it follows from the
+other conditions. -/
+def is_topological_fiber_bundle : Prop :=
+  ∀x:B, ∃e: local_homeomorph Z (B × F), ∃o:set B, x ∈ o ∧ is_open o ∧ e.source = proj⁻¹' o ∧
```

Comment:
Is there a specific reason why you don't use bundled open sets here? `∃ U : opens B, x ∈ U ... etc`

Reply:
I think `opens` are really useful when you are defining a function on open sets, for instance a sheaf. Otherwise, it doesn't really buy you anything compared to the plain set definition, and adds coercions that could get in your way, so I'd rather stick with the naive definition.

Reply:
Ok, that seems reasonable.